File: Weekly_Contests/Biweekly_99/Q3.cpp
class Solution {
public:

    int countWays(vector<vector<int>>& ranges) {
        // mind blown.....
        int res = 1, last = -1, mod = 1e9 + 7;
        sort(ranges.begin(), ranges.end());
        for (auto r: ranges) {   
            if (last < r[0])
                res = res * 2 % mod;
            last = max(last, r[1]);
        }
        return res;
    }

    int find(int x, vector<int>& root){
        if (x != root[x]){
            root[x] = find(root[x], root);
        }
        return root[x];
    }

    void do_union(int x, int y, vector<int>& root){
        int rx = find(x, root);
        int ry = find(y, root);
        if (rx != ry){
            root[ry] = rx;
        }
    }

    bool is_overlap(vector<int>& l1, vector<int>& l2){
        return l1[1] >= l2[0];
    }

    int countWays(vector<vector<int>>& ranges) {
        sort(ranges.begin(), ranges.end());
        int n = ranges.size();
        
        vector<int> root;
        for (int i = 0; i < n; i++){
            root.push_back(i);
        }
        
        int idx = 0;

        while (idx < n){
            vector<int> cur = ranges[idx];
            int j = idx + 1;
            while (j < n && is_overlap(cur, ranges[j])){
                do_union(idx, j, root);
                cur = {min(cur[0], ranges[j][0]), max(cur[1], ranges[j][1])};
                j++;
            }
            idx = j;
        }
        unordered_map<int, vector<int>> count;
        for (int i = 0; i < n; i++){
            count[root[i]].push_back(i);
        }
        // to prevent int overflow....
        int ans = 1;
        for (int i = 0; i < count.size(); i++){
            ans = (ans * 2) % int(pow(10, 9) + 7); 
        }
        return ans;

    }
};

File: Weekly_Contests/Biweekly_99/Q3.py
class Solution:
    
    def countWays(self, A: List[List[int]]) -> int:
        # mind blown....
        pre = -1
        res = 0
        for a,b in sorted(A):
            res += pre < a
            pre = max(pre, b)
        return pow(2, res, 10**9+7)



    def countWays(self, ranges: List[List[int]]) -> int:
        # Union find + interval sort + overlapping, good question!
        ranges.sort(key=lambda x:x[0])
        #print(ranges)
        n = len(ranges)
        root = [i for i in range(n)]

        idx = 0
        
        def is_overlap(l1, l2):
            return l1[1] >= l2[0]
        
        def find(x):
            if x != root[x]:
                root[x] = find(root[x])
            return root[x]
        
        def union(x, y):
            rx, ry = find(x), find(y)
            if rx != ry:
                root[ry] = rx

        while idx < n:
            cur = ranges[idx]
            j = idx + 1
            while j < n and is_overlap(cur, ranges[j]):
                union(idx, j)
                cur = [min(cur[0], ranges[j][0]), max(cur[1], ranges[j][1])]
                j += 1
            idx = j
        count = collections.defaultdict(list)
        for i in range(n):
            count[root[i]].append(i)
        #print(count)
        return (2 ** (len(count))) % (10**9 + 7)